addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
      */
     _onPaste(ev) {
         ev.preventDefault();
+
+        const sel = this.document.getSelection();
+        const cleanHtmlOptions = {
+            selectionStartInP: !!closestElement(sel.anchorNode, 'p'),
+            selectionEndInP: !!closestElement(sel.focusNode, 'p')
+        };
+
+        if (!sel.isCollapsed) {
+            this.deleteRange(sel);
+            this.historyStep();
+        }
+
         const files = getImageFiles(ev.clipboardData);
-        const clipboardHtml = ev.clipboardData.getData('text/html');
+        const htmlClipboardData = ev.clipboardData.getData('text/html');
+        const plainTextData = ev.clipboardData.getData('text/plain');
+        let cleanHtmlData = null;
+        let shouldForceDeleteFoward = false;
+
+        if (htmlClipboardData) {
+            cleanHtmlData = this._cleanHtmlClipboardData(htmlClipboardData, cleanHtmlOptions);
+
+            const paragraphRegex = /^<p>(.+)<\/p>$/;
```

Comment:
Static regex, I realize this is splitting hairs but it's best to always keep those out of functions when possible so they're not recomputed every time.